aspnetcore/blazor/webassembly-native-dependencies.md

```diff
@@ -0,0 +1,159 @@
+
+Prebuilt dependencies typically must be built using the same version of Emscripten used to build the .NET WebAssembly runtime.
+
+## Use native code
+
+Add a simple native C function to a Blazor WebAssembly app:
+
+1. Create a new Blazor WebAssembly project.
+1. Add the `WasmBuildNative` property to a property group in the app's project file (`.csproj`):
+
+   ```xml
+   <WasmBuildNative>true</WasmBuildNative>
+   ```
+
+   > [!NOTE]
+   > The `WasmBuildNative` property can also be set with the `-p|--property` option passed to the `dotnet run` command:
+   >
+   > ```dotnetcli
+   > dotnet run -p:WasmBuildNative=true
+   > ```
+
+1. Add a `Test.c` file to the project.
+1. Add a C function in `Test.c` for computing factorials:
+
+   ```c
+   int fact(int n)
+   {
+       if (n == 0) return 1;
+       return n * fact(n - 1);
+   }
+   ```
+
+1. Add a `NativeFileReference` for `Test.c` in the app's project file:
+
+   ```xml
+   <ItemGroup>
+     <NativeFileReference Include="Test.c" />
+   </ItemGroup>
+   ```
+
+1. In a Razor component (`.razor`), add a <xref:System.Runtime.InteropServices.DllImportAttribute> for the `fact` function in the generated `Test` library and call the `fact` method from .NET code in the component:
+
+   ```razor
+   @using System.Runtime.InteropServices
+
+   <p>@fact(3)</p>
+
+   @code {
+       [DllImport("Test")]
+       static extern int fact(int n);
+   }
+   ```
+
+When you build the app with the .NET WebAssembly build tools installed, the native C code is compiled and linked into the .NET WebAssembly runtime (`dotnet.wasm`). Compiling and linking may take a few minutes. After the app is built, run the app to see the rendered factorial value.
+
+> [!NOTE]
+> During the 6.0 preview release period, you may receive a build error on subsequent builds saying that the output assembly is being used by another process. This is a known issue that will be addressed for the .NET 6 general release. To workaround the issue, rebuild the project a second time.
+
+## C++ managed method callbacks
+
+Label managed methods that are passed to C++ with the `[UnmanagedCallersOnly]` attribute.
+
+The method marked with the `[UnmanagedCallersOnly]` attribute must be `static`. To call an instance method in a Razor component, pass a `GCHandle` for the instance to C++ and then pass it back to native. Alternatively, use some other method to identify the instance of the component.
+
+The method marked with `[DllImport]` must use a C# 9.0 function pointer rather than a delegate type for the callback argument.
+
+> [!NOTE]
+> For C# function pointer types in `[DllImport]` methods, use `IntPtr` in the method signature on the managed side instead of `delegate *unmanaged<int, void>`. For more information, see [[WASM] callback from native code to .NET: Parsing function pointer types in signatures is not supported (dotnet/runtime #56145)](https://github.com/dotnet/runtime/issues/56145).
+
+## Use libraries
```

Comment:
I think this section title is misleading. It says it's about using libraries, but the actual content is about *creating* libraries. People might mistakenly think that consuming a library like SkiaSharp involves doing stuff with architecture-specific directories, when they don't need to know or care about any of that.

Should the section title be changed to something like "Packaging native dependencies in a NuGet package"?